### ISSUE TITLE
[networkd] Regression in handling of socketcan interfaces after v248

### ISSUE THREAD

**Issue by splatch** [systemd/systemd#20428]
**systemd version 249 (249.3-1-manjaro)**

This is my second, if not third, attempt to get `systemd-networkd` to manage CAN interface.
I do have a gs_usb enabled device which is properly detected by kernel. When I use standard commands listed below device works.
```
ip link set can0 type can bitrate 50000 restart-ms 100
ip link set can0 up
```
Yet, with same settings placed in network file it does not work.

**Used distribution**
 > `Manjaro / Arch`

**Linux kernel version used** (`uname -a`)
 > `Linux arch 5.13.9-1-MANJARO #1 SMP PREEMPT Sun Aug 8 09:02:03 UTC 2021 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > `x86_64`

**Expected behaviour you didn't see**
 > `Managed socketcan interface`

**Unexpected behaviour you saw**
 > `Interface does not come`

**Steps to reproduce the problem**
 > 1. Get a gs_usb dongle
 > 2. Create basic networkd configuration
 > 3. Plug dongle

**Additional program output to the terminal or log subsystem illustrating the issue**

Config file `/etc/systemd/network/80-can0.network`:
```text
[Match]
Name=*can0

[CAN]
BitRate=50000
RestartSec=100ms
```

Debug output from networkd
```text
can0: State changed: failed -> linger
can0: Permanent MAC address not found for new device, continuing without: Operation not supported
can0: Link 20 added
can0: Saved original MTU 16 (min: 0, max: 4294967295)
can0: Flags change: +NOARP +ECHO
can0: link pending udev initialization...
can0: udev initialized link
can0: State changed: pending -> initialized
can0: Link state is up-to-date
can0: found matching network '/etc/systemd/network/80-can0.network'.
can0: State changed: initialized -> configuring
can0: Requested to set CAN interface configurations
can0: Requested to activate link
can0: Setting CAN interface configurations
can0: Setting bitrate = 125000 bit/s
can0: Using default sample point
can0: Setting restart = 100ms
can0: Failed to set CAN interface configurations: No such device
can0: Failed
can0: State changed: configuring -> failed
```

I am not quite sure where from the "No such device" error comes from.

**Comment by yuwata** [comment]
Does the config work with older release, e.g. v248?

**Comment by yuwata** [comment]
Fix is waiting in PR #20431.

**Comment by splatch** [comment]
I can confirm that older release did work. I believe fixes you did will let it go. I found another issue with previous version which is related to interface reconfiguration.
